samples/web/content/apprtc/js/appcontroller.js

```diff
@@ -64,45 +66,65 @@ var AppController = function(params) {
   this.fullscreenIconSet_ = new AppController.IconSet_(
       UI_CONSTANTS.fullscreenOnSvg, UI_CONSTANTS.fullscreenOffSvg);
 
-  this.params_ = params;
-
-  this.call_ = new Call(params);
-  this.infoBox_ =
-      new InfoBox($(UI_CONSTANTS.infoDiv), this.remoteVideo_, this.call_);
-
-  this.transitionToWaitingTimer_ = null;
-
-  var roomErrors = params.errorMessages;
-  if (roomErrors.length > 0) {
-    for (var i = 0; i < roomErrors.length; ++i) {
-      this.infoBox_.pushErrorMessage(roomErrors[i]);
-    }
-    return;
+  this.loadingParams_ = loadingParams;
+  var paramsPromise = Promise.resolve(JSON.stringify({}));
+  if (this.loadingParams_.paramsFunction)
+  {
+    // Fetch params from server.
+    paramsPromise = this.loadingParams_.paramsFunction();
   }
 
-  // TODO(jiayl): replace callbacks with events.
-  this.call_.onremotehangup = this.onRemoteHangup_.bind(this);
-  this.call_.onremotesdpset = this.onRemoteSdpSet_.bind(this);
-  this.call_.onremotestreamadded = this.onRemoteStreamAdded_.bind(this);
-  this.call_.onlocalstreamadded = this.onLocalStreamAdded_.bind(this);
-
-  this.call_.onsignalingstatechange =
-      this.infoBox_.updateInfoDiv.bind(this.infoBox_);
-  this.call_.oniceconnectionstatechange =
-      this.infoBox_.updateInfoDiv.bind(this.infoBox_);
-  this.call_.onnewicecandidate =
-      this.infoBox_.recordIceCandidateTypes.bind(this.infoBox_);
+  Promise.resolve(paramsPromise).then(function(newParams) {
+    // Merge newly retrieved params with loadingParams
+    if (newParams) {
+      Object.keys(newParams).forEach(function(key) {
+        this.loadingParams_[key] = newParams[key];
+      }.bind(this));
+    }
+  }.bind(this)).catch(function(error) {
+    trace('Initializing; error getting params from server: ' + error.message);
+  }.bind(this)).then(function() {
+    // Proceed with call set up.
```

Comment:
I think it will be easier to read if you combine the two .then callbacks into a helper method.